File: examples/cpp/union_find.yosupo.test.cpp
#define PROBLEM "https://judge.yosupo.jp/problem/unionfind"
#include <iostream>
#include "examples/cpp/union_find.hpp"
#include "examples/cpp/macros.hpp"
using namespace std;

int main() {
    int n, q; cin >> n >> q;
    union_find uf(n);
    REP (i, q) {
        int t, u, v; cin >> t >> u >> v;
        if (t == 0) {
            uf.unite_trees(u, v);
        } else if (t == 1) {
            cout << uf.is_same(u, v) << endl;
        }
    }
    return 0;
}

File: examples/python/union_find_yosupo.test.py
# verification-helper: PROBLEM https://judge.yosupo.jp/problem/unionfind
import sys
input = sys.stdin.buffer.readline

from examples.python.union_find import UnionFind


def main() -> None:
    N, Q = map(int, input().split())
    uf = UnionFind(N)
    for _ in range(Q):
        t, u, v = map(int, input().split())
        if t == 0:
            uf.unite(u, v)
        else:
            print(int(uf.is_same(u, v)))


if __name__ == "__main__":
    main()
